### PR TITLE
POST /signups endpoint

### LINKED ISSUE #685: Add new Signups endpoint for Quicksilver POST
Refs #676 -- Add new `signups` route instead of posting to `chatbot`

cc @sergii-tkachenko

### PR DESCRIPTION
**Pull request by aaronschachter**
#### What's this PR do?
- Creates `app/routes` directory to maintain app routes. This could easily end up as a `controllers` directory instead, as we slowly refactor / kill off `CampaignBotController` (and hopefully `DonorsChooseBotController` #648) into Mongoose static/instance methods per model classes (refs #682, #686)
- Refactors the POST `/v1/chatbot` with `signup_id`  and `signup_source` as new `/v1/signups` route, expecting `id` and `source` params, removes to `routes/signups`
- Moves Campaigns routes (#677) to `routes/campaigns`, Chatbot routes to `routes/chatbot` 
- Refactors `router-chatbot.postMobileCommonsProfile` function as a User instance method `postMobileCommonsProfileUpdate`  to easily share between `routes/signups` and `routes/chatbot`
- Renames `api` directory as `app`. Named it as `api` initially to mimic Sails and Quicksilver API, but `app` to work better for new directories and related reading
- Returns a 422 error if incoming Signup is for a Campaign that isn't a CampaignBot Campaign (refs #642)
#### How should this be reviewed?

Verify posting to `/v1/signups` works as expected, as well as all other endpoints
#### Any background context you want to provide?

Related reading re: best practices for Express strucure: 
- https://github.com/terlici/base-express
- http://www.codekitchen.ca/guide-to-structuring-and-building-a-restful-api-using-express-4

Would love to get second opinions on how to structure (include `app` directory or just define directories like `models`, `controllers`/`routes` in project root)?
#### Relevant tickets

Fixes #685
#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Tested on staging.

### CODE REVIEW

File: app/routes/signups.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    .then((user) => {
+      const scope = req;
+      scope.user = user;
+      scope.signup = currentSignup;
+      scope.campaign = app.locals.campaigns[currentSignup.campaign];
+      const controller = app.locals.controllers.campaignBot;
+      const msg = controller.renderResponseMessage(scope, 'menu_signedup_external');
+      user.postMobileCommonsProfileUpdate(req, process.env.MOBILECOMMONS_OIP_CHATBOT, msg);
+
+      return res.send(msg);
```
Comment (aaronschachter) [inline app/routes/signups.js]: @sergii-tkachenko What should the response look like for success? This is just copy we send as the confirmation message if success, although seems to make sense to return an object... just not sure what it should contain. Same question for what happens if POST `/signups` receives a Signup for a Campaign that's not featured as CampaignBot Campaign in L22

File: documentation/endpoints/signups.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+Name | Type | Description
+--- | --- | ---
+`x-gambit-api-key` | `string` | **Required.** 
+ 
+**Input**
+
+Name | Type | Description
+--- | --- | ---
+`id` | `number` |  **Required.**  Signup ID to update Gambit cache for
```
Comment (aaronschachter) [inline documentation/endpoints/signups.md]: @sergii-tkachenko Also ok to pass `id` and `source` to POST `signups` now (seemed redundant to keep the params as `signup_id` and `signup_source` for new `signups` endpoint.

File: app/routes/signups.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const express = require('express');
+const router = express.Router(); // eslint-disable-line new-cap
```
Comment (sergii-tkachenko) [inline app/routes/signups.js]: I found new-cap rule not very effective. Might make sense to just turn it off for properties for good:
https://github.com/DoSomething/quicksilver-api/blob/b69f8c90fa1901cfc1fb250262236071e56a43b9/test/.eslintrc#L8-L10

File: app/routes/signups.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      currentSignup = signup;
+      if (!app.locals.campaigns[signup.campaign]) {
+        throw new CampaignNotFoundError();
+      }
+
+      return app.locals.db.users.lookup('id', currentSignup.user);
+    })
+    .then((user) => {
+      if (!user) {
+        return res.status(500).send({ error: 'Cannot find user for signup' });
```
Comment (sergii-tkachenko) [inline app/routes/signups.js]: Wouldn't it be passed to next `.then((user)`? Maybe throw an error instead?

File: app/routes/signups.js (hunk 1 above)
Comment (sergii-tkachenko) [inline app/routes/signups.js]: Something like [this](https://github.com/DoSomething/northstar/blob/dev/documentation/endpoints/auth.md#revoke-token)?

```js
// 200 OK

{
  "success": {
    "code": 200,
    "message": "That refresh token has been successfully revoked."
  }
}
```

Comment (sergii-tkachenko) [inline app/routes/signups.js]: W/o message is fine as well.

File: app/routes/signups.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      scope.signup = currentSignup;
+      scope.campaign = app.locals.campaigns[currentSignup.campaign];
+      const controller = app.locals.controllers.campaignBot;
+      const msg = controller.renderResponseMessage(scope, 'menu_signedup_external');
+      user.postMobileCommonsProfileUpdate(req, process.env.MOBILECOMMONS_OIP_CHATBOT, msg);
+
+      return res.send(msg);
+    })
+    .catch(CampaignNotFoundError, () => {
+      const msg = `Campaign ${currentSignup.campaign} is not a CampaignBot campaign.`;
```
Comment (sergii-tkachenko) [inline app/routes/signups.js]: Just note for future: `CampaignNotFoundError` could generate this message and recommended HTTP status code.

File: documentation/endpoints/signups.md (hunk 2 above)
Comment (sergii-tkachenko) [inline documentation/endpoints/signups.md]: Yes, that's definitely better.

**Review by sergiitk — CHANGES_REQUESTED** [review]

File: app/models/User.js [hunk 6]
```diff
@@ -121,6 +122,24 @@ userSchema.statics.createForMobileCommonsRequest = function (req) {
 };
 
 /**
+ * Updates user's Mobile Commons Profile to send msgTxt as SMS via target mobileCommonsOIP.
+ */
+userSchema.methods.postMobileCommonsProfileUpdate = function (req, mobileCommonsOIP, msgTxt) {
```
Comment (aaronschachter) [inline app/models/User.js]: Shouldn't space be here? Linter isn't complaining...

File: app/routes/signups.js (hunk 4 above)
Comment (aaronschachter) [inline app/routes/signups.js]: Will add this in next PR.

**Pull request merged**